src/runtime_src/xdp/profile/database/static_info_database.cpp

```diff
@@ -2640,31 +2648,39 @@ namespace xdp {
         aie::readAIEMetadata(data.first, data.second, aieMetadata);
     }
 
-    if (!metadataReader)
+    if (!metadataReader) {
       xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT",
                               "AIE metadata read failed!");
-    else
+    } else {
       xrt_core::message::send(xrt_core::message::severity_level::debug, "XRT",
                               "AIE metadata read successfully!");
+      std::lock_guard<std::mutex> lock(aieMetadataReaderLock) ;
+      metadataReaders.emplace(deviceId, std::move(metadataReader));
+    }
   }
 
   const xdp::aie::BaseFiletypeImpl*
-  VPStaticDatabase::getAIEmetadataReader() const
+  VPStaticDatabase::getAIEmetadataReader(uint64_t deviceId) const
   {
+    std::lock_guard<std::mutex> lock(aieMetadataReaderLock) ;
     xrt_core::message::send(xrt_core::message::severity_level::info, "XRT", "AIE metadataReader requested");
-    return metadataReader.get();
+    if (metadataReaders.find(deviceId) == metadataReaders.end())
+      return nullptr;
+
+    return metadataReaders.at(deviceId).get();
   }
 
   void VPStaticDatabase::setAIEGeneration(uint64_t deviceId) {
     std::lock_guard<std::mutex> lock(deviceLock) ;
     if (deviceInfo.find(deviceId) == deviceInfo.end())
       return;
 
-    if (!metadataReader)
+    if (!getAIEmetadataReader(deviceId))
       return;
 
     try {
-      auto hwGen = metadataReader->getHardwareGeneration();
+      std::lock_guard<std::mutex> lock(aieMetadataReaderLock) ;
```

Comment:
We are already fetching metadataReader for "deviceId" at Line 2678. Please reuse that and remove the aiemetadatareader lock here.